finality-aleph/src/new_network/connection_manager/connections.rs

```diff
@@ -0,0 +1,85 @@
+use crate::{new_network::PeerId, SessionId};
+use std::collections::{HashMap, HashSet};
+
+/// Keeps track of connections we should maintain taking into account data from many sessions.
+pub struct Connections {
+    associated_sessions: HashMap<PeerId, HashSet<SessionId>>,
+    peers_by_session: HashMap<SessionId, HashSet<PeerId>>,
+}
+
+impl Connections {
+    /// Creates a new object, initially without any connections.
+    pub fn new() -> Self {
+        Connections {
+            associated_sessions: HashMap::new(),
+            peers_by_session: HashMap::new(),
+        }
+    }
+
+    /// Mark the specified peers as ones we should be connected to for the given session.
+    pub fn add_peers(&mut self, session_id: SessionId, peers: HashSet<PeerId>) {
+        for peer in peers {
+            self.associated_sessions
+                .entry(peer)
+                .or_insert_with(HashSet::new)
+                .insert(session_id);
+            self.peers_by_session
+                .entry(session_id)
+                .or_insert_with(HashSet::new)
+                .insert(peer);
+        }
+    }
+
+    /// Assume we no longer need to be connected to peers from the given session.
+    /// Returns the peers we no longer have any reason to be connected to.
+    pub fn remove_session(&mut self, session_id: SessionId) -> HashSet<PeerId> {
+        let mut result = HashSet::new();
+        if let Some(peers) = self.peers_by_session.remove(&session_id) {
+            for peer in peers {
+                if let Some(mut sessions) = self.associated_sessions.remove(&peer) {
+                    sessions.remove(&session_id);
+                    if !sessions.is_empty() {
+                        self.associated_sessions.insert(peer, sessions);
+                    } else {
+                        result.insert(peer);
+                    }
+                }
+            }
+        }
+        result
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::Connections;
+    use crate::{new_network::PeerId, SessionId};
+    use sc_network::PeerId as ScPeerId;
+    use std::collections::HashSet;
+
+    fn random_peer_ids(num: usize) -> HashSet<PeerId> {
+        (0..num).map(|_| PeerId(ScPeerId::random())).collect()
+    }
+
+    #[test]
+    fn removes_peer_after_single_session() {
```

Comment:
I hate to be that pettifogging, but you wrote unit tests, that are assuming correctness of functions `add_peers` and `remove_session` without testing them separately

Reply:
They cannot be tested separately using the public interface, without inspecting the implementation details of the struct. The only behaviour I can test is by calling them in various configurations and looking at the outputs of the latter. This is also the contract that this struct is supposed to fulfill, so I don't consider this a problem.

"Pettifogging" is a brilliant word (and approach ;3), I did not know that one.